### PR TITLE
Turn Symfony App into API mode and create Player Entity

### PR DESCRIPTION
**Pull request by Clebiez**
[Trello Card](https://trello.com/c/6eNXBkyi/1-075-as-brice-i-want-to-find-an-opponent-to-play-with)

- [x] Create a /login route which return a player by a given name or create one instead.
- [x] Update game entity to handle Player instead of hash.
- [x] Make 3 API routes : POST /games, POST /games/id/join and GET /games/id to allow players to join a game  
- [x] First player is chosen randomly
image
image
image

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -49,6 +49,7 @@
         "twig/twig": "^2.12|^3.0"
     },
     "require-dev": {
+        "squizlabs/php_codesniffer": "*",
```
Comment (alexisjanvier) [inline composer.json]: what a weird version number

File: src/Controller/GameController.php [hunk 2]
```diff
@@ -48,113 +52,89 @@ public function notify(PublisherInterface $publisher, UuidV4 $gameId, array $par
...
 
-        $response = $this->redirectToRoute(
-            'game',
-            [
-                "id" => $game->getId()
-            ]
-        );
-        $response->headers->setCookie(new Cookie($this::COOKIE_KEY, $playerHash));
-
+        $response = new JsonResponse(new GameDto($game));
```
Comment (alexisjanvier) [inline src/Controller/GameController.php:76]: I'm not a big fan that ... I find that using a method like `toJson` would make the code more explicit.

File: src/Controller/GameController.php [hunk 3]
```diff
@@ -48,113 +52,89 @@ public function notify(PublisherInterface $publisher, UuidV4 $gameId, array $par
...
+        $response->headers->set('Access-Control-Allow-Origin', '*');
```
Comment (alexisjanvier) [inline src/Controller/GameController.php]: this could certainly be refactorized prior to the action

File: src/Controller/GameController.php [hunk 4]
```diff
@@ -48,113 +52,89 @@ public function notify(PublisherInterface $publisher, UuidV4 $gameId, array $par
...
     /**
-     * @Route("/games/{id}/join", name="joinGame")
+     * @Route("/games/{id}/join", name="game_join", methods={"POST"})
      */
     public function joinGame(Request $request, string $id, PublisherInterface $publisher): Response
     {
-        $playerHash = $request->cookies->get($this::COOKIE_KEY);
+        $content = $request->toArray();
+        if (!$content || !$content["playerId"]) {
+            return new JsonResponse("Id and playerId are required", 400);
```
Comment (alexisjanvier) [inline src/Controller/GameController.php:87]: This is in line with the previous comment on CORS. Here, as you return an answer, won't you miss the CORS?
In my opinion, CORS should be treated in a more generic way.

File: src/Controller/GameController.php (hunk 2 above)
Comment (alexisjanvier) [reply src/Controller/GameController.php:76]: ok, `toJson` is not a solution !

File: src/Dto/GameDto.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace App\Dto;
+
+use App\Entity\Game;
+
+class GameDto
```
Comment (alexisjanvier) [inline src/Dto/GameDto.php:7]: I'm not sure I understand what this dto is for ... the `get` methods were not public ?

File: src/Entity/Game.php [hunk 6]
```diff
@@ -24,14 +24,14 @@ class Game
     private $id;
 
     /**
-     * @ORM\Column(type="string", length=255, nullable=true)
+     * @ORM\ManyToOne(targetEntity="Player", cascade={"all"}, fetch="EAGER")
```
Comment (alexisjanvier) [inline src/Entity/Game.php:27]: What is the `fetch="EAGER"` for?

File: src/Repository/PlayerRepository.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+{
+    public function __construct(ManagerRegistry $registry)
+    {
+        parent::__construct($registry, Player::class);
+    }
+
+    // /**
+    //  * @return Player[] Returns an array of Player objects
+    //  */
+    /*
```
Comment (alexisjanvier) [inline src/Repository/PlayerRepository.php]: commented code spoted

**Review by alexisjanvier — CHANGES_REQUESTED** [review]

File: src/Repository/PlayerRepository.php (hunk 7 above)
Comment (Clebiez) [reply src/Repository/PlayerRepository.php]: Yes it's the Symfony Make bundle who harass me with comments :D

File: src/Entity/Game.php (hunk 6 above)
Comment (Clebiez) [reply src/Entity/Game.php:27]: `EAGER` mean when I fetch a Game entity, `player1` will be directly initialized as Player entity.

Other can be `Lazy` and means that Doctirne will hydrate `player1` when I will call `$game>getPlayer1()`.
Second options should improve performance if I don't use everytime all relationships.

But If I always use this relationships, `Eager` is more indicated.

File: src/Dto/GameDto.php (hunk 5 above)
Comment (Clebiez) [reply src/Dto/GameDto.php:7]: If I don't use something like a srializer, the response will be empty : `{}`

File: src/Controller/GameController.php (hunk 3 above)
Comment (Clebiez) [reply src/Controller/GameController.php]: Fixed by adding NelmioCorsBundle

File: .env [hunk 8]
```diff
@@ -27,3 +27,7 @@ APP_SECRET=ea92ba9f5d776bb49c40d8ddb959703c
 # IMPORTANT: You MUST configure your server version, either here or in config/packages/doctrine.yaml
 #
 
+
+###> nelmio/cors-bundle ###
+CORS_ALLOW_ORIGIN='^http://(localhost|127\.0\.0\.1)(:[0-9]+)?$'
```
Comment (alexisjanvier) [inline .env:32]: :+1:

**Review by alexisjanvier — APPROVED** [review]

**Pull request merged**